### PR TITLE
[WIP] make some implementation classes private or final

### PR DESCRIPTION
**Pull request by mkeskells**
to promote discussion, and limit future binary compatability issues

### CODE REVIEW

**Comment by ShaneDelmore** [comment]
I’d like to see a deprecation phase here.

File: src/library/scala/collection/immutable/ArraySeq.scala [hunk 1]
```diff
@@ -36,7 +36,7 @@ sealed abstract class ArraySeq[+A]
     * the expected immutability. Its element type does not have to be equal to the element type of this ArraySeq.
     * A primitive ArraySeq can be backed by an array of boxed values and a reference ArraySeq can be backed by an
     * array of a supertype or subtype of the element type. */
-  def unsafeArray: Array[_]
+  private[immutable] def unsafeArray: Array[_]
```
Comment (szeiger) [inline src/library/scala/collection/immutable/ArraySeq.scala:39]: `immutable.ArraySeq` relies on convention to keep the array contents immutable. We need it for performance reasons, in particular for varargs. Depending on what course we take for varargs (https://github.com/scala/bug/issues/11040) we could consider a safer approach if we don't need it for varargs anymore.

File: src/library/scala/collection/IndexedSeqView.scala [hunk 2]
```diff
@@ -37,53 +37,53 @@ object IndexedSeqView {
   type SomeIndexedSeqOps[A] = IndexedSeqOps[A, AnyConstr, _]
 
   @SerialVersionUID(3L)
-  class Id[+A](underlying: SomeIndexedSeqOps[A])
+  private class Id[+A](underlying: SomeIndexedSeqOps[A])
```
Comment (szeiger) [inline src/library/scala/collection/IndexedSeqView.scala:40]: All View implementation classes need to be public. You have to use them directly when you implement a collection that falls outside of the standard four collection kinds. (for example, search for `View.` in `AnyRefMap` or `s.c.BitSet`)

File: src/library/scala/collection/immutable/BitSet.scala [hunk 3]
```diff
@@ -113,7 +113,7 @@ object BitSet extends SpecificIterableFactory[Int, BitSet] {
     else new BitSetN(elems)
   }
 
-  class BitSet1(val elems: Long) extends BitSet {
+  private class BitSet1(val elems: Long) extends BitSet {
```
Comment (szeiger) [inline src/library/scala/collection/immutable/BitSet.scala:116]: Making these classes private appears reasonable but we should go through a deprecation cycle. I don't think they are subclassed anywhere so they could also be make final (after deprecation).

File: src/library/scala/collection/concurrent/TrieMap.scala [hunk 4]
```diff
@@ -987,7 +987,7 @@ object TrieMap extends MapFactory[TrieMap] {
 
   val inodeupdater = AtomicReferenceFieldUpdater.newUpdater(classOf[INodeBase[_, _]], classOf[MainNode[_, _]], "mainnode")
 
-  class MangledHashing[K] extends Hashing[K] {
+  private class MangledHashing[K] extends Hashing[K] {
```
Comment (szeiger) [inline src/library/scala/collection/concurrent/TrieMap.scala:990]: Makes sense (since `TrieMap` is final anyway) but should be deprecated first.

**Comment by dwijnand** [comment]
Looks like [`@deprecatedInheritance`](https://www.scala-lang.org/api/current/scala/deprecatedInheritance.html) could be used here.

File: src/library/scala/collection/Iterator.scala [hunk 5]
```diff
@@ -144,7 +144,7 @@ trait Iterator[+A] extends IterableOnce[A] with IterableOnceOps[A, Iterator, Ite
    *
    *  Typical uses can be achieved via methods `grouped` and `sliding`.
    */
-  class GroupedIterator[B >: A](self: Iterator[B], size: Int, step: Int) extends AbstractIterator[immutable.Seq[B]] {
+  class GroupedIterator[B >: A]private[Iterator] (self: Iterator[B], size: Int, step: Int) extends AbstractIterator[immutable.Seq[B]] {
```
Comment (szeiger) [inline src/library/scala/collection/Iterator.scala:147]: I can't think of a compelling use case for calling the constructor directly but we're exposing the type anyway. Hiding the constructor wouldn't really help.

File: src/library/scala/collection/immutable/Range.scala [hunk 6]
```diff
@@ -479,7 +479,7 @@ object Range {
 
   // As there is no appealing default step size for not-really-integral ranges,
   // we offer a partially constructed object.
-  class Partial[T, U](private val f: T => U) extends AnyVal {
+  private[scala] class Partial[T, U](private val f: T => U) extends AnyVal {
```
Comment (szeiger) [inline src/library/scala/collection/immutable/Range.scala:482]: This needs to be public. Numeric type implementations must be able to call and return `Partial`. In my experience having `private[scala]` on an element that is not directly in package `scala` is often a bad idea. If we can't restrict its usage to its own package, user code may want to call it, too.

**Comment by lrytz** [comment]
@mkeskells is this still on your radar?

**Comment by mkeskells** [comment]
I will have a look when I get back. Travelling atm. Probably will see what is viable and pop it in a new pr. Suggestions welcome

Thanks for reminding me

**Comment by mkeskells** [comment]
This PR was to promote discussion, which has started, and some of the changes are made via other PRs
Closing this now as we are finalising the API for 2.13.0 and there is nothing here that should block that

**Pull request closed without merging**